### ISSUE TITLE
`Bijector` support for batch computation

### ISSUE THREAD

**Issue by torfjelde** [TuringLang/Bijectors.jl#35]
## Problem
Currently the `Bijector` interface doesn't "officially" support batch computations. With this I mean, what do you do when a `SimplexBijector` receives an `AbstractMatrix` rather than an `AbstractVector` as is expected? In some cases it's fairly straight forward to do so, e.g. the `SimplexBijector` only makes sense when applied to a vector so when we receive a matrix we know it's a batch and we can just do the right thing. In other cases what to do with a `Real` vs. `AbstractVector{<:Real}` is not so clear, e.g. `Exp` _elementwise_ is a well-defined bijection on `Real`, `AbstractVector`, etc. so what do we do in `logabsdetjac(b::Exp, x)` when `x isa AbstractArray`? Is `x` a batch of inputs so we should return `[y[i] for i = 1:length(x)]`, or is it intended as a vector and we should return `sum(y)` as is need for the distribution `MvLogNormal`?

To drive the point home, what do we do in this situation:
```julia
julia> struct Scale{T}
           a::T
       end

julia> (b::Scale)(x) = b.a .* x

julia> b = Scale(2.0)
Scale{Float64}(2.0)

julia> x = randn()
0.2414826706733002

julia> b(x)
0.4829653413466004

julia> logabsdetjac(b::Scale, x) = log(abs(b.a))
logabsdetjac (generic function with 1 method)

julia> logabsdetjac(b, x)
0.6931471805599453

julia> x = rand(2) # suppose we want to scale vectors elementwise instead
2-element Array{Float64,1}:
 0.7910530695591607 
 0.03333662406026572

julia> b(x) # transform still works nicely
2-element Array{Float64,1}:
 1.5821061391183213 
 0.06667324812053144

julia> logabsdetjac(b, x)
0.6931471805599453

julia> x = rand(2) # batch of 2
2-element Array{Float64,1}:
 0.546411899713563  
 0.16872866221657734

julia> b(x) # transform works on batch as expected
2-element Array{Float64,1}:
 1.092823799427126 
 0.3374573244331547

julia> logabsdetjac(b, x) # NO! Want this to return [logabsdetjac(b, x[1]), logabsdetjac(b, x[2])]!
0.6931471805599453
```
In this example we _could_ use the parametric type `T` of `Scale{T}` to decide what to do, but what do we do when `T <: AbstractVector` and `logabsdetjac` is given an `AbstractMatrix`? Is `Scale` intended to act on a matrix by scaling each column elementwise or is it intended to act on a vector and this matrix is actually a batch?

At the time of writing, some of the bijectors support batch computation by "accident", e.g. `SimplexBijector` there is only one thing to do when given an `AbstractMatrix`, but it's crucial that we have support for such a feature for _all_ bijectors for it to be of any use. If not, a lot of incompatibilities will arise when users try to compose two bijectors, one supporting batch-computation and one does not; then you end up with size-mismatches in the `logabsdetjac` field.

### 1. `Univariate`, `Multivariate`, etc.
We could introduce something similar to Distributions.jl's `Univariate`, `Multivariate`, etc. The example of `Scale` would then instead be
```julia
struct Scale{O, T}
    a::T
end

(b::Scale)(x) = b.a .* x

logabsdetjac(b::Scale{Univariate}, x::Real) = log(abs(b.a))
logabsdetjac(b::Scale{Univariate}, x::AbstractVector) = log(abs(b.a)) .* ones(length(x))

logabsdetjac(b::Scale{Multivariate}, x::AbstractVector{<:Real}) = sum(log(abs(b.a)))
logabsdetjac(b::Scale{Multivariate}, x::AbstractMatrix{<:Real}) = sum(log(abs(b.a))) .* ones(size(x, 2))
```
and so on.

As suggested by @xukai92, this can also be used to specify the device location if we just use aliases like `Univariate = Real`, etc. This way one could have specific implementations for `CuArray`, etc.

The main issue I see is that we'll then have to force this upon all bijectors, some for which it is redundant information as they are only well-defined on a single specific space.

### 2. Fixed-size
A super-simple solution is to force each bijector should to the dimensions of the output upon construction. I personally don't like it as it can be very inflexible and in _most_ cases the bijection is well-defined on a particular space, so the dimension can be quite redundant.

### 3. ???

**Comment by torfjelde** [comment]
So I think there are two main approaches to solving this, and I'd love to hear what people think.

## Size
Explicitly provide size of expected input for _all_ bijectors you construct (you also need to do it for the ones which are only well-defined on a specific type of input, e.g. `SimplexBijector`).
```julia
abstract type Bijector{Size} end

# Example of construction:
# Expects input `x` with `size(x) = (3, )` 
# OR `size(x) = (3, n)` where `n` is the number of inputs ("batch-computation")
Logit{(3, )}()

# Can also make it a bit nicer by adding constructors, e.g.:
Logit(size::Tuple) = Logit{size}()
```
### Pros
- Catch _size_ discrepancies for compositions of bijectors during composition (compile-time) rather than during evaluation (run-time)
- Can do checks at compile-time when using `StaticArray` (but whether you're catching size discrepancies when actually _evaluting_ a `Bijector` on an input at compile-time or run-time doesn't really matter as far as I can tell)
- The upcoming `Stacked` bijector in #36 provides a way of stacking 0- and 1-dimensional bijectors to act on a single vector, e.g. apply `Logit` to `x[1:3]` and `SimplexBijector` to `x[4:end]` or something. With static sizes we'd be able to provide sane defaults to the ranges, e.g. stacking `Logit{(N1, )}` and `SimplexBijector{(N2, )}` will apply `Logit` to `x[1:N1]` and `SimplexBijector` to `x[N1 + 1:end]`. Can also do this for more than just two bijectors. At the moment, the user has to provide the ranges unless they're all 0-dimensional in which case we'll just do elementwise application.

### Cons
- It puts a lot of effort on the user. You need to do this explicitly for _all_ bijectors.
- When working with non-static arrays you'll in many cases have to reconstruct the `Bijector` on each input to ensure dimensionality of the `Bijector` is correct. This can be inefficient (and annoying) in certain cases.
- It can also get fairly ugly internally, which might also be a problem for the user if they want to implement their own `Bijector`

## Dimensionality
Instead of the `size` of the expected input we can parameterize `Bijector` by the dimensionality of the expected input, i.e. `length(size)`.

```julia
abstract type Bijector{N} end

# Example of construction:
Logit{1}() # expects vector input
```

In a lot of cases a bijector is only well-defined for a particular dimensionality, e.g. `SimplexBijector` is only well-defined on a 1D vector. Such cases will then subtype `Bijector{specific_value_of_N}`, e.g.
```julia
struct SimplexBijector <: Bijector{1} end
```

### Pros
- Catching _dimensionality_ (not _size) discrepancies at composition rather than at evaluation
- From a user-perspective, there's not much change. Only certain bijectors require the user to provide the dimensionality of the expected input since most are only well-defined on, say, a 1D vector.
- Often easy to provide sane defaults, e.g. `Log() = Log{0}()`.
- Provide default constructor for `Stacked` in the case where the bijectors to stack are all 0-dimensional

### Cons
- For a small subset of bijectors, the user has to specify the dimensions.

**Comment by torfjelde** [comment]
Personally I prefer the "Dimensionality" approach :man_shrugging:

**Comment by xukai92** [comment]
Seems that for the `Stacked` purposes the "Size" approach is better?

**Comment by torfjelde** [comment]
Yeah, but I'm also afraid that `Size` will cause issues with type-stability due to needing run-time information for construction (unless using `StaticArray`, which is worse than standard `Array` in higher dimensions). So for `Stacked` it would indeed be better in the case where we construct the bijector at the top-level, but as soon as we construct it inside a function call or whatever we'd fall back to dynamic dispatch, right? This wouldn't be the case if we only depend on the dimensionality of the input array (which is known at compile time). 

I think I really prefer not using `Size` now. Alternatively we can provide a `StaticBijector` or something later on? E.g. Introduce `AbstractBijector`, let `abstract Bijector{N} <: AbstractBijector` and `abstract StaticBijector{Size} <: AbstractBijector`?

**Comment by torfjelde** [comment]
I think I'm actually quite fond of using the "Dimensionality" approach now, and then we can introduce `StaticBijector` later on as follows:

```julia
# Replaces `Bijector` as root-type
abstract type AbstractBijector end

# Redefine `Bijector{N}` as subtype of `AbstractBijector`
# => all existing code remains the same
abstract type Bijector{N} <: AbstractBijector end

# Introduce new type parameterized by `Size`
abstract type StaticBijector{Size} <: AbstractBijector end

# implement fixed-size implementations of different bijectors
...
```

**Comment by mohamed82008** [comment]
I agree with Tor on this one. Here is an excerpt from Slack:
> No encoding the size in the type is not good imo. It will bite us later. If the user passes x::StaticArray, we can specialize on the size of x but there is no need for the bijector to be limited to a certain size imo.
>
> I am not sure checking size-discrepancy between bijectors in Composed upon construction is the right thing either. For instance the bijector may be a linear map with the map matrix size being runtime info.
>
> Or a scale which makes sense regardless of the dimension.

**Comment by mohamed82008** [comment]
`StaticBijector` also sounds good to me.

**Comment by xukai92** [comment]
I like the `StaticBijector` idea!

**Comment by torfjelde** [comment]
> I like the StaticBijector idea!

Great! So we go for the "Dimensionality" approach now, and then we can introduce `StaticBijector` later on? I think `StaticBijector` requires some more thought and work, so personally I don't want to rush that.

**Comment by xukai92** [comment]
Yes it sounds like a good plan to me.

**Comment by torfjelde** [comment]
I think this was mistakenly closed by the merging of PR #36 ? It's still unsolved, but an implementation and thus a fix of this issue is almost done and can be found in PR #44

**Comment by torfjelde** [comment]
Just to make sure; everyone is happy with this approach?

If so, I'll make the final changes to PR #44 so we can get this merged. I have a lot of work on my personal github which uses this branch as a base and it's been working really well this far.

**Comment by torfjelde** [comment]
Should we close this issue now that #44 has been merged? Or keep it up for future discussion? I'm still not 100% convinced the approach we've taken is absolutely optimal (though at the moment it seems like the best route)

**Comment by willtebbutt** [comment]
I'm pro- closing this. Could always open a new issue at a later date if someone has a bright idea, but I'm in agreement that this solution feels pretty optimal for now.

**Comment by torfjelde** [comment]
We should also bump the version number at some point, since this is possibly a breaking change.
But there's a lot of work ready to go based on #44 being merged, so should probably get that into master before making an actual releases

**Comment by torfjelde** [comment]
Closed by #44
